src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +474,361 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
                 iEndIndex = endIndex
             };
 
-            return expansion.InsertExpansion(textSpan, textSpan, this, LanguageServiceGuid, out ExpansionSession) == VSConstants.S_OK;
+            if (expansion.InsertExpansion(textSpan, textSpan, this, LanguageServiceGuid, out _state._expansionSession) == VSConstants.S_OK)
+            {
+                // This expansion is not derived from a symbol, so make sure the state isn't tracking any symbol
+                // information
+                _state.ClearSymbolInformation();
+                return true;
+            }
+
+            if (!(SubjectBuffer.GetFeatureOnOffOption(CompletionOptions.EnableArgumentCompletionSnippets) ?? false))
+            {
+                // Argument completion snippets are not enabled
+                return false;
+            }
+
+            var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document is null)
+            {
+                // Couldn't identify the current document
+                return false;
+            }
+
+            var symbols = ThreadingContext.JoinableTaskFactory.Run(() => GetReferencedSymbolsToLeftOfCaretAsync(document, caretPosition: triggerSpan.End, cancellationToken));
+
+            var methodSymbols = symbols.OfType<IMethodSymbol>().ToImmutableArray();
+            if (methodSymbols.Any())
+            {
+                var methodName = dataBufferSpan.GetText();
+                var snippet = CreateMethodCallSnippet(methodName, includeMethod: true, ImmutableArray<IParameterSymbol>.Empty, cancellationToken);
+
+                var doc = new DOMDocumentClass();
+                if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+                {
+                    _state._methods = methodSymbols;
+                    _state._method = null;
+
+                    if (expansion.InsertSpecificExpansion(doc, textSpan, this, LanguageServiceGuid, pszRelativePath: null, out _state._expansionSession) == VSConstants.S_OK)
+                    {
+                        Debug.Assert(_state._methods == methodSymbols);
+                        Debug.Assert(_state._method == null);
+
+                        if (_signatureHelpControllerProvider.GetController(TextView, SubjectBuffer) is { } controller)
+                        {
+                            // Register a handler for ModelUpdated. To avoid the possibility of more than one
+                            // handler being in the list, remove any current one before adding the new one. This
+                            // handler is only changed on the main thread, and the event is only invoked on the main
+                            // thread, so additional synchronization is not necessary to prevent lost events.
+                            controller.ModelUpdated -= OnModelUpdated;
+                            controller.ModelUpdated += OnModelUpdated;
```

Comment:
Do we eventually unsubscribe from this somewhere? I get it's probably safe in that the OnModelUpdated code will probably bail, but it seems safer just in case this feature goes sideways and the user turns it off?

Reply:
> Do we eventually unsubscribe from this somewhere?

No, the controller and the client are tied to a view, so they have the same lifetime. Closing the window will release both.

Reply:
I don't disagree with that analysis; but I'm not sure I like our code not having a comment saying that's the intent versus this being an oversight. My strong preference would be to ensure this code is correct on it's own without making a cross-feature analysis that way.

Reply:
Rewrote the event listener hookup/cleanup.